source/microbit/persistent.c

```diff
@@ -0,0 +1,176 @@
+        address += (len>>2)<<2;
+        data += (len>>2)<<2;
+        len &= 3;
+    }
+    if (len) {
+        nrf_nvmc_write_bytes((uint32_t)address, (const uint8_t *)data, len);
+    }
+    while (NRF_NVMC->READY == NVMC_READY_READY_Busy);
+    ticker_start();
+}
+
+static inline bool can_write(const int8_t *dest, const int8_t *src, uint32_t len) {
+    const int8_t *end = dest + len;
+    while (dest < end) {
+        if (((~(*dest)) & *src) != 0) {
+            DEBUG(("PERSISTENCE DEBUG: %d bytes from %x need to be erased\r\n", len, dest));
+            return false;
+        }
+        dest++;
+        src++;
+    }
+    DEBUG(("PERSISTENCE DEBUG: %d bytes from %x do not need to be erased\r\n", len, dest));
+    return true;
+}
+
+STATIC void erase_page(const void *page) {
+    DEBUG(("PERSISTENCE DEBUG: Erasing page %x\r\n", page));
+    nrf_nvmc_page_erase((uint32_t)page);
+}
+
+bool is_persistent_page_aligned(const void *ptr) {
+    return (((uint32_t)ptr) & (persistent_page_size()-1)) == 0;
+}
+
+void persistent_erase(const void *dst, uint32_t len) {
+    DEBUG(("PERSISTENCE DEBUG: Erase persistent %d bytes at %x\r\n", len, dst));
+    const int8_t *dest = dst;
+    const int8_t *end_erase = dest+len;
+    const int8_t *addr = dest;
+    const uint32_t page_size = persistent_page_size();
+    if (is_persistent_page_aligned(dst) && len == page_size) {
+        erase_page(dst);
+        return;
+    }
+    int8_t *page = (void *)(((uint32_t)dest)&(-page_size));
+    int8_t *tmp_storage = NULL;
+    while (addr < end_erase) {
+        int8_t *next_page = page + page_size;
+        uint32_t in_page = min(end_erase-addr, next_page-dest);
+        if (in_page < page_size) {
+            if (tmp_storage == NULL) {
+                tmp_storage = gc_alloc(page_size, false);
+            }
+            memcpy(tmp_storage, page, page_size);
+            memset(tmp_storage+(dest-page), -1, in_page);
+        }
+        erase_page(page);
+        if (in_page < page_size) {
+            persistent_write_unchecked(page, tmp_storage, page_size);
+        }
+        addr += in_page;
+    }
+}
+
+void persistent_write(const void *dst, const void *src, uint32_t len) {
+    DEBUG(("PERSISTENCE DEBUG: Write persistent %d bytes from %x to %x\r\n", len, src, dst));
+    const int8_t *dest = dst;
+    const int8_t *addr = src;
+    const int8_t *end_data = src+len;
+    const uint32_t page_size = persistent_page_size();
+    int8_t *page = (void *)(((uint32_t)dest)&(-page_size));
+    int8_t *tmp_storage = NULL;
+    while (addr < end_data) {
+        int8_t *next_page = page + page_size;
+        uint32_t data_in_page = min(end_data-addr, next_page-dest);
+        if (can_write(dest, addr, data_in_page)) {
+            persistent_write_unchecked(dest, addr, data_in_page);
+        } else {
+            if (tmp_storage == NULL) {
+                tmp_storage = gc_alloc(page_size, false);
```

Comment:
This is kind of bad.  It means you need 1k of free heap (that's the nRF51's page size, right?) to do file writing.  What if the allocation fails?  Also, should use m_new(uint8_t, page_size).